### PR TITLE
IVT Microsoft.VisualStudio.LanguageServerClient.Razor

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Specifically razor will soon need `IAsynchronousOperationListenerProvider` and it's descendants.

### CODE REVIEW

**Review by sharwell — CHANGES_REQUESTED** [review]
These types will need to be provided through types in an external access assembly, and not directly through IVTs. For example, `IApexAsynchronousOperationListenerProviderAccessor` provides Apex with access to the APIs it needs from `IAsynchronousOperationListenerProvider`. You'll want a `IRazorAsynchronousOperationListenerProviderAccessor` to provide the APIs needed by Razor.

**Comment by ryanbrandenburg** [comment]
@sharwell the ExternalAccess pattern is a new one to me, clean if a little verbose. I assume the idea there just to put hard and explicit limits on which APIs get IVT'd?

**Comment by sharwell** [comment]
> I assume the idea there just to put hard and explicit limits on which APIs get IVT'd?

It helps in two ways:

1. We know which partners are using which internal APIs
2. If we change an internal signature, we can update the wrapper to avoid breaking the partner (e.g. if we add a parameter, we can update the wrapper to pass a default value for that parameter instead of forcing the partner to update their use in a coordinated insertion)

File: src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+#nullable disable
+
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal interface IRazorAsynchronousOperationListenerProviderAccessor
+    {
+        IAsynchronousOperationListener GetListener(string featureName);
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs:13]:
```suggestion
        RazorAsynchronousOperationListenerWrapper GetListener(string featureName);
```

Comment (sharwell) [reply src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs:13]: Here's another wrapper:

https://github.com/dotnet/roslyn/blob/master/src/Tools/ExternalAccess/Razor/Remote/RazorRemoteServiceCallbackIdWrapper.cs

You won't need the data contract attributes in this case, but the wrapper concept would be the same.

File: src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs [hunk 2]
```diff
@@ -4,12 +4,10 @@
...
 #nullable disable
 
-using Microsoft.CodeAnalysis.Shared.TestHooks;
-
 namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
 {
     internal interface IRazorAsynchronousOperationListenerProviderAccessor
     {
-        IAsynchronousOperationListener GetListener(string featureName);
+        IRazorAsynchronousOperationListener GetListener(string featureName);
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs]: 💡 Use of an interface adds allocation overhead to this wrapper. Switching to a `readonly struct` will eliminate the overhead.

File: src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerWrapper.cs [hunk 3]
```diff
@@ -8,18 +8,16 @@
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
 {
-    internal class RazorAsynchronousOperationListener
+    internal class RazorAsynchronousOperationListenerWrapper
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerWrapper.cs:11]:
```suggestion
    internal readonly struct RazorAsynchronousOperationListenerWrapper
```

**Comment by ryanbrandenburg** [comment]
\azp run

File: src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/IRazorAsynchronousOperationListenerProviderAccessor.cs:5]:
```suggestion
```

File: src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerProviderAccessor.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    [Export(typeof(IRazorAsynchronousOperationListenerProviderAccessor))]
+    [Shared]
+    internal sealed class RazorAsynchronousOperationListenerProviderAccessor : IRazorAsynchronousOperationListenerProviderAccessor
+    {
+        private readonly AsynchronousOperationListenerProvider _implementation;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RazorAsynchronousOperationListenerProviderAccessor(
+            AsynchronousOperationListenerProvider implementation)
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerProviderAccessor.cs:21]:
```suggestion
            IAsynchronousOperationListenerProvider implementation)
```

**Review by sharwell — APPROVED** [review]

File: src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerProviderAccessor.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    [Export(typeof(IRazorAsynchronousOperationListenerProviderAccessor))]
+    [Shared]
+    internal sealed class RazorAsynchronousOperationListenerProviderAccessor : IRazorAsynchronousOperationListenerProviderAccessor
+    {
+        private readonly AsynchronousOperationListenerProvider _implementation;
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorAsynchronousOperationListenerProviderAccessor.cs:16]:
```suggestion
        private readonly IAsynchronousOperationListenerProvider _implementation;
```

**Pull request merged**